drivers/sensor/akm09918c/akm09918c_emul.c

```diff
@@ -130,14 +133,63 @@ static int akm09918c_emul_init(const struct emul *target, const struct device *p
 	return 0;
 }
 
+static int akm09918c_emul_backend_set_channel(const struct emul *target, enum sensor_channel ch,
+					      q31_t value, uint8_t shift)
+{
+	if (!target || !target->data) {
+		return -EINVAL;
+	}
+
+	struct akm09918c_emul_data *data = target->data;
+	uint8_t reg;
+
+	switch (ch) {
+	case SENSOR_CHAN_MAGN_X:
+		reg = AKM09918C_REG_HXL;
+		break;
+	case SENSOR_CHAN_MAGN_Y:
+		reg = AKM09918C_REG_HYL;
+		break;
+	case SENSOR_CHAN_MAGN_Z:
+		reg = AKM09918C_REG_HZL;
+		break;
+	/* This function only supports setting single channels, so skip MAGN_XYZ */
+	default:
+		return -ENOTSUP;
+	}
+
+	/* Set the ST1 register to show we have data */
+	data->reg[AKM09918C_REG_ST1] |= AKM09918C_ST1_DRDY;
+
+	/* Convert fixed-point microgauss (uGs) value into a register value. The smallest increment
+	 * is 1500 uGs, so shift away the fractional portion of the value and consider only the
+	 * integer portion. Limit the value to the supported range.
+	 */
+
+	int16_t reg_val = CLAMP(arithmetic_shift_right(value, shift),
```

Comment:
`arithmetic_shift_right` doesn't work here, you'll need to check if shift is negative and basically shift left if positive, shift right if negative. Since this is a test, you're welcome to use the zDSP library for this (the scale function will handle this for you) or you can manually do the check `shift < 0 ? (value >> -shift) : (value << shift)`